### PR TITLE
Windows: Run containerd as managed process

### PR DESCRIPTION
**Pull request by olljanat**
**- What I did**
Continued from where #42089 left so now users can switch to containerd by just adding
```json
{
  "default-runtime": "io.containerd.runhcs.v1"
}
```
to `daemon.config` without need separately install or configure containerd.

Needs that containerd binaries are included to static binaries installation package #43892

Also need to be backported to 22.06.x if this approach is accepted.

**- How I did it**
Copied logic from Linux implementation and updated. In additionally used non-default named pipe so it is possible to that both Docker with containerd and separate containerd instance on same machine.

**- How to verify it**
CI was modified on way that it run Windows 2019 + containerd tests like before and Windows 2022 + containerd tests using this new approach.

Additionally manual tests can be done by
1. Add containerd binaries to same folder with Docker binaries.
2. Add daemon.json configuration
3. Start with `dockerd.exe -D` and check that you see `Starting daemon managed containerd` message.

**- Description for the changelog**
 

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by olljanat** [comment]
One open questions related to this one is that if we should only look `containerd.exe` from folder where Docker binaries are instead of PATH? That would be then different approach than what Linux is using but maybe still good idea as people might run containerd + nerdctl on same machines.

File: .github/workflows/windows.yml [hunk 1]
```diff
@@ -346,10 +346,14 @@ jobs:
         name: Starting test daemon
         run: |
           Write-Host "Creating service"
-          If ("${{ matrix.runtime }}" -eq "containerd") {
+          If ("${{ matrix.runtime }}" -eq "containerd" -and "${{ matrix.os }}" -eq "windows-2019") {
             $runtimeArg="--containerd=\\.\pipe\containerd-containerd"
```
Comment (TBBle) [inline .github/workflows/windows.yml:350]: Now that the `remote.Debug.Address` no longer defaults to the system containerd, should we also pass the system container's debug socket on this CLI? I haven't looked to see how that socket is used, this is just flagging from the change in _libcontainerd/supervisor/remote_daemon_windows.go_.

Comment (TBBle) [reply .github/workflows/windows.yml:350]: I see the Unix-side code doesn't do this either, perhaps this debug socket is only used for a locally-started containerd?

File: cmd/dockerd/daemon_windows.go [hunk 2]
```diff
@@ -93,7 +94,30 @@ func newCgroupParent(config *config.Config) string {
 	return ""
 }
 
-func (cli *DaemonCli) initContainerD(_ context.Context) (func(time.Duration) error, error) {
+func (cli *DaemonCli) initContainerD(ctx context.Context) (func(time.Duration) error, error) {
+	if cli.Config.ContainerdAddr == "" && cli.Config.DefaultRuntime == config.WindowsV2RuntimeName {
+		cli.Config.ContainerdAddr = supervisor.GrpcPipeName
```
Comment (TBBle) [inline cmd/dockerd/daemon_windows.go:99]: I'll note that the Unix-side implementation attempts to detect a running system containerd first; this is feasible in Windows from I've read on Stack Overflow, but probably not _as_ trivial as the Unix-side `os.Lstat`. That might be a nice improvement for the next iteration, as then neither test path would require passing in the containerd address.

File: .github/workflows/windows.yml [hunk 3]
```diff
@@ -323,7 +323,7 @@ jobs:
           }
       -
         name: Starting containerd
-        if: matrix.runtime == 'containerd'
+        if: matrix.runtime == 'containerd' && matrix.os == 'windows-2019'
```
Comment (TBBle) [inline .github/workflows/windows.yml:326]: I don't know GitHub Actions that well, but would it be clearer to have the `runtime` axis with three values (`builtin`, `managed-containerd`, `system-containerd`) and also flag at that point to skip running `windows-2019/system-containerd` and `windows-2022/managed-containerd`? That will also make it easier to adjust for "Windows 2025" for example. (Playing the long game. ^_^)

Comment (TBBle) [reply .github/workflows/windows.yml:326]: Ah yes, there's [an `exclude` parameter for the matrix definition](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#excluding-matrix-configurations)..

File: cmd/dockerd/daemon_windows.go (hunk 2 above)
Comment (olljanat) [reply cmd/dockerd/daemon_windows.go:99]: Yes I noticed that. However there containerd was taken to use on very early phase.

Windows containerd is already in used with nerdctl and Kubernetes on some environments and if we want to avoid issue which you mentioned in https://github.com/moby/moby/issues/43892#issuecomment-1200334488
> A specific expectation/benefit for me of this is that the zip includes the exact containerd version which was used in CI, reducing the scope for mysterious bugs.

also on those environments it might be better to use this approach.

File: .github/workflows/windows.yml (hunk 1 above)
Comment (olljanat) [reply .github/workflows/windows.yml:350]: Found this comment from containerd code `DefaultDebugAddress is the default winpipe address for pprof data` so I understand that it purely for debugging but I have no idea how to use it.

File: cmd/dockerd/daemon_windows.go (hunk 2 above)
Comment (TBBle) [reply cmd/dockerd/daemon_windows.go:99]: In that comment, I was thinking more about "static" binary downloads, which I think have a fairly focused use-case of an all-in-one download with no external dependencies or installers.

Personally I'm undecided whether I would expect a bundled-containerd in, the successor of DockerMSFTProvider, or I'd expect it to depend on another DockerContainerdProvider. (I guess a winget setup will be the likely successor, which implies someone builds an MSI for this, and so they are making the decisions or extending the options anyway). I spent a long time in Debian where the answer would be "Of course they should be separate packages", but I recognise that this may be a somewhat-fringe feeling in the real world. ^_^

_Personally_ I like to be able to mix-and-match versions in my own processes, but lock down versions in the outputs. But I have the luxury of time to do that. If I just needed something working _now_, I'd just grab one of the versions of containerd which is recorded as "tested against" in the docs or CI of the dockerd I'm grabbing.

File: .github/workflows/windows.yml (hunk 1 above)
Comment (TBBle) [reply .github/workflows/windows.yml:350]: I don't have the source in front of me, but GitHub search is telling me that Docker doesn't ever _use_ the socket. Either I'm just not searching well enough, or some refactoring has disappeared the use-cases for this and no one noticed? Or it's actually supposed to be used by someone running `ctr pprof` or another tool entirely, not by Docker itself.

Anyway, still browsing GitHub, Docker's libcontainerd/supervisor always passes in a Debug socket (pointing to a private socket in the changes you've made here parallel to the private containerd socket), and when we start containerd in the system test, without passing one in, it doesn't start listening (i.e. default is "" on Windows, Linux and Solaris).

So what we have now is fine, since we don't try and collect the pprof data in the test run, and we shouldn't see any conflicts between system-containerd and docker-private containerd instances over the socket name.

Comment (TBBle) [reply .github/workflows/windows.yml:350]: There might be a slight separate bug related to this, as [`monitorDaemon` deletes the GRPC socket, but not the Debug socket](https://github.com/moby/moby/blob/master/libcontainerd/supervisor/remote_daemon.go#L278) when starting/restarting containerd. Probably not an issue, but if we start seeing bundled-containerd startups/restarts failing because the pprof socket already exists, then that'll be why. (Not sure if npipe semantics on Windows are different from unix sockets in this regard, my recollection is that neither _requires_ explicit removal after the ends are closed, so removing the GRPC socket in that code might just be bonus safety).

File: cmd/dockerd/daemon_windows.go (hunk 2 above)
Comment (olljanat) [reply cmd/dockerd/daemon_windows.go:99]: > Personally I like to be able to mix-and-match versions in my own processes, but lock down versions in the outputs.

Yea, I have noticed that you looks to have a lot of time for stuff which is nice. I don't but I still would like to see that Windows + containerd combination to be usable enough that people can start using it on dev environments and provide feedback.

> I dunno if there's many other cases like that around which expect a given host to have only one container runtime instance (used as a sharing point between k8s and Docker/BuildKit)

There is no much to share nowadays as Docker uses only graphdriver (at least Windows implementation, not sure what is status with migrating Linux parts) and containerd uses snapshotter so those cannot share images. Look https://blog.mobyproject.org/where-are-containerds-graph-drivers-145fc9b7255

File: .github/workflows/windows.yml (hunk 1 above)
Comment (TBBle) [reply .github/workflows/windows.yml:350]: And then there's #43901.

File: .github/workflows/windows.yml (hunk 3 above)
Comment (crazy-max) [reply .github/workflows/windows.yml:326]: Using `include` pattern in the matrix to run specific against specific fields would also work.

Comment (olljanat) [reply .github/workflows/windows.yml:326]: Ok. However I think that there is other open questions here and https://github.com/moby/moby/issues/43892 to be answered first before I will continue this further.

**Comment by thaJeztah** [comment]
Quick blurb from discussing this in the maintainers meeting; there were some discussions recently about wether or not we want to continue having the "containerd as child process" option, or if we should deprecate it, so I wanted to check with other maintainers if we'd be good with this PR.

- After some discussion, we think it's ok to continue having this as a feature, as there's practical use-cases (mostly for "non-prod")
- For Windows, it's definitely easier than to set up a service
- We may want to explore options for users to create their own `containerd.toml` config file that can be used for the "managed" containerd, so that we don't have to add more code in the daemon to allow setting those config options.


TL;DR; generally, the features added in this PR is ok (but we didn't review the code) 👍

**Comment by thaJeztah** [comment]
> We may want to explore options for users to create their own containerd.toml config file that can be used for the "managed" containerd, so that we don't have to add more code in the daemon to allow setting those config options.

I opened https://github.com/moby/moby/pull/43946 to implement that suggestion

Let me try giving this PR a look soon (I'll also do a quick try if (where) it conflicts with https://github.com/moby/moby/pull/43946)

File: cmd/dockerd/daemon_windows.go [hunk 4]
```diff
@@ -93,7 +94,30 @@ func newCgroupParent(config *config.Config) string {
 	return ""
 }
 
-func (cli *DaemonCli) initContainerD(_ context.Context) (func(time.Duration) error, error) {
+func (cli *DaemonCli) initContainerD(ctx context.Context) (func(time.Duration) error, error) {
+	if cli.Config.ContainerdAddr == "" && cli.Config.DefaultRuntime == config.WindowsV2RuntimeName {
```
Comment (thaJeztah) [inline cmd/dockerd/daemon_windows.go:98]: Heads up that I'm changing this function to `initContainerd` (lowercase) in https://github.com/moby/moby/pull/43945. In that PR I also removed the `.Config` from these, to make it _slightly_ less verbose.

Perhaps this should use an early return here?

```go
if cli.ContainerdAddr != "" {
	// use system containerd at the given address.
	return nil, nil
}

if cli.Config.DefaultRuntime != config.WindowsV2RuntimeName {
	// containerd is only used if the V2 (io.containerd.runhcs.v1) runtime is used.
	return nil, nil
}
```

As to the `config.WindowsV2RuntimeName` check; I guess we currently use that as switch to toggle between containerd and the built-in runtime, so makes sense; slightly wondering if (in light of https://github.com/moby/moby/pull/43946), we might run into issues if alternative runtime may come to exist (not really urgent I guess).

File: cmd/dockerd/daemon_windows.go (hunk 2 above)
Comment (thaJeztah) [inline cmd/dockerd/daemon_windows.go:99]: Curious about the;

```go
cli.Config.ContainerdAddr = supervisor.GrpcPipeName
```

Wouldn't that be set as part of `setDefaults()` ? https://github.com/moby/moby/blob/c8fc98937d7bb75093e3eaeb0d9fd6b717e46bdc/libcontainerd/supervisor/remote_daemon_windows.go#L14

Or was this specifically for the `initContainerdRuntime()`?

File: cmd/dockerd/daemon_windows.go [hunk 5]
```diff
@@ -93,7 +94,30 @@ func newCgroupParent(config *config.Config) string {
...
+		cli.Config.ContainerdAddr = supervisor.GrpcPipeName
+		system.InitContainerdRuntime(cli.Config.ContainerdAddr)
```
Comment (thaJeztah) [inline cmd/dockerd/daemon_windows.go:100]: ^^ related to my previous comment; if we set it as part of `setDefaults()`, then this could be moved to the end of this function, and use `r.Address()`


With that, we also wouldn't have to export `supervisor.GrpcPipeName`

File: cmd/dockerd/daemon_windows.go [hunk 6]
```diff
@@ -93,7 +94,30 @@ func newCgroupParent(config *config.Config) string {
...
+
+		var waitForShutdown func(time.Duration) error
+		logrus.Debug("Starting daemon managed containerd")
```
Comment (thaJeztah) [inline cmd/dockerd/daemon_windows.go:103]: In https://github.com/moby/moby/pull/43945, I changed this to

- `managed containerd` (removed "daemon"), and made the logs all lowercase
- changed `Debug` to `Info` (for consistency with some other things we start when starting the daemon)

Perhaps something like;

```go
logrus.Info("starting managed containerd")
```

Or we could even use the same log;

```go
logrus.Info("containerd not running, starting managed containerd")
```

File: cmd/dockerd/daemon_windows.go [hunk 7]
```diff
@@ -93,7 +94,30 @@ func newCgroupParent(config *config.Config) string {
...
+		opts, err := cli.getContainerdDaemonOpts()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to generate containerd options")
+		}
+
+		r, err := supervisor.Start(ctx, filepath.Join(cli.Config.Root, "containerd"), filepath.Join(cli.Config.ExecRoot, "containerd"), opts...)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to start containerd")
+		}
+		logrus.Debug("Started daemon managed containerd")
```
Comment (thaJeztah) [inline cmd/dockerd/daemon_windows.go:113]: While working on https://github.com/moby/moby/pull/43945, I noticed that this log was redundant (`remote. startContainerd()` already logs when it's started, including `PID`, and `address`), so this line can be removed.

File: cmd/dockerd/daemon_windows.go [hunk 8]
```diff
@@ -93,7 +94,30 @@ func newCgroupParent(config *config.Config) string {
...
+		cli.Config.ContainerdAddr = r.Address()
+
+		// Try to wait for containerd to shutdown
+		waitForShutdown = r.WaitTimeout
+		return waitForShutdown, nil
```
Comment (thaJeztah) [inline cmd/dockerd/daemon_windows.go:118]: `waitForShutdown` is only used here, so this could be inlined (and the var further up removed);

```go
return r.WaitTimeout, nil
```

Comment (thaJeztah) [reply cmd/dockerd/daemon_windows.go:118]: With my suggestions applied, this function would look like;

```go
func (cli *DaemonCli) initContainerd(ctx context.Context) (func(time.Duration) error, error) {
	if cli.ContainerdAddr != "" {
		// use system containerd at the given address.
		system.InitContainerdRuntime(cli.ContainerdAddr)
		return nil, nil
	}

	if cli.DefaultRuntime != config.WindowsV2RuntimeName {
		return nil, nil
	}

	logrus.Info("containerd not running, starting managed containerd")
	opts, err := cli.getContainerdDaemonOpts()
	if err != nil {
		return nil, errors.Wrap(err, "failed to generate containerd options")
	}

	r, err := supervisor.Start(ctx, filepath.Join(cli.Root, "containerd"), filepath.Join(cli.ExecRoot, "containerd"), opts...)
	if err != nil {
		return nil, errors.Wrap(err, "failed to start containerd")
	}
	cli.ContainerdAddr = r.Address()
	system.InitContainerdRuntime(r.Address())

	// Try to wait for containerd to shutdown
	return r.WaitTimeout, nil
}
```

And the diff between Windows and Linux getting really small (maybe even up to a level that we could abstract some of the platform specifics);

Screenshot 2022-08-11 at 16 48 16

File: libcontainerd/supervisor/remote_daemon_windows.go [hunk 9]
```diff
@@ -7,13 +7,14 @@ import (
 )
 
 const (
-	grpcPipeName  = `\\.\pipe\containerd-containerd`
-	debugPipeName = `\\.\pipe\containerd-debug`
+	binaryName    = "containerd.exe"
+	GrpcPipeName  = `\\.\pipe\docker-containerd`
```
Comment (thaJeztah) [inline libcontainerd/supervisor/remote_daemon_windows.go:11]: See my other comment; possibly we don't need to export this one

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! I left some thoughts/ramblings 😅 - happy to hear your thoughts.

I also had a quick try locally to rebase this PR on top of https://github.com/moby/moby/pull/43945, which only gave a small conflict, so not troublesome.

File: daemon/config/config_windows.go [hunk 10]
```diff
@@ -8,8 +8,16 @@ const (
 	// This is used by the `default-runtime` flag in dockerd as the default value.
 	// On windows we'd prefer to keep this empty so the value is auto-detected based on other options.
 	StockRuntimeName = ""
```
Comment (thaJeztah) [inline daemon/config/config_windows.go:10]: Looking at this bit again, and wondering if we should continue "not using" this for Windows, or if we should put `com.docker.hcsshim.v1` here (default (for now) to be the built-in runtime, and `io.containerd.runhcs.v1` to be the "opt in" for containerd). Where at some point we can change the default.

Comment (thaJeztah) [reply daemon/config/config_windows.go:10]: Comment was added in https://github.com/moby/moby/commit/7ccf750daad37aa12e790d5f057b7ae6a140e0b8 (https://github.com/moby/moby/pull/42089) ./cc @cpuguy83 perhaps you recall the details on this 🤔

**Comment by thaJeztah** [comment]
@olljanat I rebased my local branch and added the suggestions I made above (including splitting that one block to a separate commit); if useful, you can reset your branch to that one; https://github.com/moby/moby/compare/master..thaJeztah:windows_containerd_as_proces_rebase_and_suggestions?expand=1

(also happy to push to your PR here; let me know)

**Comment by olljanat** [comment]
> (also happy to push to your PR here; let me know)

@thaJeztah please do so. I will try stay away from keyboard next three weeks (after tomorrow) and most probably don't have time to push those changes in place before that.

**Comment by thaJeztah** [comment]
Assuming that's for Holiday/Vacation, enjoy! 🌴

**Comment by thaJeztah** [comment]
oh! let me try to update this PR branch; hold on

**Comment by thaJeztah** [comment]
done 👍

**Comment by olljanat** [comment]
> oh! let me try to update this PR branch; hold on

Oh, I forgot that you can update others PRs.

**Comment by thaJeztah** [comment]
Yes! I don't always use it, but in this case it's probably good to preserve the discussion 👍 

I also pushed https://github.com/moby/moby/pull/43951 with the bit that I removed from this one (and the part that may need further discussion)

**Review by ericbarrionuevo — APPROVED** [review]

**Comment by olljanat** [comment]
> Yes! I don't always use it, but in this case it's probably good to preserve the discussion 👍

Yea. Btw I rebased this now without changes as there was conflicting changes in `.github/workflows/.windows.yml`

**Comment by olljanat** [comment]
@thaJeztah Should this be included and backported to 23.0 or was there something unclear still?

**Comment by neersighted** [comment]
Discussed in today's maintainer's call: This is something we'd like to see, but we don't want to rush it for 23.0.0.

The talking points in short are:
- Would like to see as much shared with Linux as possible.
- Needs to detect an already running containerd (try to talk to the named pipe?) and not start another instance.
  - The only distribution of Moby that uses containerd on Windows, MCR, would be broken by this (Docker Desktop and Docker CE do not presently use containerd on Windows).
- Too big a change in behavior given this hasn't seen a lot of review recently (and of the latest commits/the current design); it needs more time for a proper review.

**Comment by olljanat** [comment]
> The only distribution of Moby that uses containerd on Windows, MCR, would be broken by this 

  @neersighted dockerd.exe installed as part of Mirantis Container Runtime uses containerd already? Since when? Working code which supports that does not exist on Moby before 23.0.0 ( #42164 #41479 #42089 #43866 ) and there is multiple known issues which why lot of tests are disabled on that combination #41455

So I'm quite sure that only reason why MCR installs containerd is that because nowadays Kubernetes uses that directly without Docker. Reason to rush this one to 23.0.0 is finally provide first version of where users can start easily pilot containerd configuration.

**Comment by neersighted** [comment]
Right, we're not actively running containers using the runhcs runtime (that lives on a branch and will likely not be supported commercially for some time) -- what I mean to say is that unexpected things will happen for users based on @thaJeztah's summary of the state of this PR (I have not reviewed it).

It's only one of the bullet points as well -- the main takeaway is that the risk seems relatively high, the maintainers who have reviewed this in the past aren't confident in the state today, and we'd rather include this in a patch release as an experimental/power-user feature vs. delay the release of 23.0.0 or rush it in half-baked to a release that has had multiple RCs already.

**Comment by neersighted** [comment]
Bump -- now that 23.0.0 is out, hopefully we'll have some bandwidth to review this. How do you feel about the idea of factoring out the common Linux code into a (partially) shared implementation in this PR?

**Comment by CharityKathure** [comment]
- [x] Merge master into windows-containerd-as-proces
- [x] Resolve Merge conflicts
- [x] factor out the common Linux code into a (partially) shared implementation
- [ ]  Address failing checks

**Comment by olljanat** [comment]
@CharityKathure I think that is easier if you just open new pull request which replaces this. Then you can create it based on latest version of body in moby/moby so you will get rid of all those merge conflicts.

**Comment by thaJeztah** [comment]
- superseded by https://github.com/moby/moby/pull/47955
